### ISSUE TITLE
f2py generated C code doesn't recognize "long_long" 

### ISSUE THREAD

**Issue by LukasAdamowicz** [numpy/numpy#15095]
When using a .f2py_f2cmap to specify 64 bit integers (int64 from iso_fortran_env) as "long_long", the code will not compile if any of the subroutine parameters are specified as int64 (as long_long appears in the generated c-code for the function).  

If using a local variable with the int64 kind the error does not occur.

Generating the C code, modifying "long_long" to "long long" does allow compilation.

### Reproducing code example:

test.f95
```fortran
      subroutine func1(n, x, res)
        use, intrinsic :: iso_fortran_env, only: int64, real64
        implicit none
        integer(int64), intent(in) :: n
        real(real64), intent(in) :: x(n)
        real(real64), intent(out) :: res
Cf2py   intent(hide) :: n

        res = sum(x)
      end
```

.f2py_f2cmap
```
dict(real=dict(real32='float', real64='double'), integer=dict(int64='long_long'))
```

Compiling with GCC (11.0.0)/GFortran(9.2.0) on MacOS 10.15.1:
```shell script
f2py -c -m test test.95
```

### Error message:
```shell script
.../testmodule.c:141:33: error: unknown type name 'long_long'
extern void F_FUNC(func1,FUNC1)(long_long*,double*,double*);
                                ^
.../testmodule.c:163:46: error: unknown type name 'long_long'
                           void (*f2py_func)(long_long*,double*,double*)) {
                                             ^
.../testmodule.c:168:3: error: use of undeclared identifier 'long_long'
  long_long n = 0;
```

### Numpy/Python version information:

- Numpy: 1.17.4
- Python: 3.7.5

**Comment by WarrenWeckesser** [comment]
> When using a .f2py_f2cmap to specify 64 bit integers (int64 from iso_fortran_env) as "long_long", 

Do you have a reference to documentation or other source that explains why you would do this?  `long_long` is not a valid C identifier.

**Comment by WarrenWeckesser** [comment]
Hmmm... maybe this *should* work.  The default `f2cmap_all` in [the source](https://github.com/numpy/numpy/blob/175ba9a7bde4e00202c8cb117b84693fd8a7bfd2/numpy/f2py/capi_maps.py#L162) uses `'long_long'`.   Unfortunately, the use of `.f2py_f2cmap` [is not documented](https://github.com/numpy/numpy/issues/15096).   If anyone digs into the f2py source to figure out how this is supposed to work, please also update the f2py documentation at https://github.com/numpy/numpy/tree/master/doc/source/f2py!

**Comment by LukasAdamowicz** [comment]
I initially assumed it would be "long long" as well, but if you try "long long" instead, when generating the files the process specifically ignores those integer values, as the `f2cmap` reader doesn't like that value, and instead requests "long_long"

**Comment by LukasAdamowicz** [comment]
After realizing that I had some code where the functionality __did__ work, the intermediate C module file has this at the top:

```C
/*********************** See f2py2e/cfuncs.py: typedefs ***********************/
#ifdef _WIN32
typedef __int64 long_long;
#else
typedef long long long_long;
typedef unsigned long long unsigned_long_long;
#endif
```

This is not present in other files (below). The only significant difference in this file is that it has 2 include statements at the top (for the kind=8 version of FFTPACK)

```C
/*********************** See f2py2e/cfuncs.py: typedefs ***********************/
/*need_typedefs*/
```

**Comment by mattip** [comment]
The f2cmap was documented in gh-15113 and gh-15106. Does [that documentation](https://numpy.org/devdocs/f2py/advanced.html?highlight=f2cmap#dealing-with-kind-specifiers) provide enough information to fix this issue?

**Comment by pv** [comment]
@mattip I think no, because setting `long_long` as explained in that documentation doesn't seem to work always.

**Comment by HaoZeke** [comment]
Thanks @pv! The issue here is that even though the map was being modified internally, the necessary helper functions were not being generated. The fix is then straightforward and in #21712.

```bash
f2py -m testy test.f95 -c
python -c "import numpy as np; import testy; print(testy.func1(np.ones(3)))"
3.0
```
